Contention scheduling actions in HashedWheelTimerScheduler

**Version Information**
Version of Akka.NET? 1.5.18
Which Akka.NET Modules? Akka.Core

**Describe the bug**
It seems like the dotnet 6 and greater version of the HashedWheelTimerScheduler introduces significant contention/blocking when registering many actions. We have thousands of actors who schedule reminders to stop themselves, and often this happens at the same time. When I target dotnet 5 we don't see this issue. We initially noticed this because our thread pool count went through the roof.

See this gist:
https://gist.github.com/raypurchasett/605ee8a0beeeafa6834b9daa6c960a8e

We've seen this running on: windows (rider), docker desktop (windows), and EKS.

@raypurchasett so @Arkatufus was able to reproduce your result - I wrote a little script to do this as well.

## NET Core App 3.1

```text
C:\Program Files\dotnet\sdk\8.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(32,5): warning NETSDK1138: The target framework 'netcoreapp3.1' is out of support and will not receive security updates in the future. Please refer to https://aka.ms/dotnet-core-support for more information about the support policy. [E:\Repositories\repros\HighSchedulerCpu\HighSchedulerCpu.csproj::TargetFramework=netcoreapp3.1]
C:\Program Files\dotnet\sdk\8.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(32,5): warning NETSDK1138: The target framework 'netcoreapp3.1' is out of support and will not receive security updates in the future. Please refer to https://aka.ms/dotnet-core-support for more information about the support policy. [E:\Repositories\repros\HighSchedulerCpu\HighSchedulerCpu.csproj::TargetFramework=netcoreapp3.1]
E:\Repositories\repros\HighSchedulerCpu\Program.cs(29,9): warning CS8602: Dereference of a possibly null reference. [E:\Repositories\repros\HighSchedulerCpu\HighSchedulerCpu.csproj::TargetFramework=netcoreapp3.1]
Waiting to start
Starting
[INFO][04/11/2024 02:45:33.369Z][Thread 0009][akka://System/user/1] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0005][akka://System/user/0] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0007][akka://System/user/3] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0016][akka://System/user/11] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0017][akka://System/user/12] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0013][akka://System/user/8] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0019][akka://System/user/9] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0010][akka://System/user/7] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0006][akka://System/user/2] 10
[INFO][04/11/2024 02:45:33.369Z][Thread 0012][akka://System/user/6] 10
[INFO][04/11/2024 02:45:33.370Z][Thread 0008][akka://System/user/4] 10
[INFO][04/11/2024 02:45:33.370Z][Thread 0011][akka://System/user/5] 10
[INFO][04/11/2024 02:45:33.370Z][Thread 0018][akka://System/user/13] 10
[INFO][04/11/2024 02:45:33.370Z][Thread 0014][akka://System/user/14] 10
[INFO][04/11/2024 02:45:33.370Z][Thread 0015][akka://System/user/10] 10
[INFO][04/11/2024 02:45:33.415Z][Thread 0012][akka://System/user/20] 0
```

## NET 8.0

```text
C:\Program Files\dotnet\sdk\8.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(32,5): warning NETSDK1138: The target framework 'netcoreapp3.1' is out of support and will not receive security updates in the future. Please refer to https://aka.ms/dotnet-core-support for more information about the support policy. [E:\Repositories\repros\HighSchedulerCpu\HighSchedulerCpu.csproj::TargetFramework=netcoreapp3.1]
Waiting to start
Starting
[INFO][04/11/2024 02:44:59.214Z][Thread 0010][akka://System/user/3] 6345
[INFO][04/11/2024 02:44:59.214Z][Thread 0009][akka://System/user/1] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0024][akka://System/user/6] 4052
[INFO][04/11/2024 02:44:59.214Z][Thread 0021][akka://System/user/39] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0029][akka://System/user/8] 0
[INFO][04/11/2024 02:44:59.214Z][Thread 0017][akka://System/user/37] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0011][akka://System/user/2] 6345
[INFO][04/11/2024 02:44:59.214Z][Thread 0013][akka://System/user/5] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0018][akka://System/user/164] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0020][akka://System/user/87] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0022][akka://System/user/40] 5574
[INFO][04/11/2024 02:44:59.214Z][Thread 0016][akka://System/user/36] 6345
[INFO][04/11/2024 02:44:59.214Z][Thread 0026][akka://System/user/88] 2023
[INFO][04/11/2024 02:44:59.214Z][Thread 0007][akka://System/user/0] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0008][akka://System/user/34] 6345
[INFO][04/11/2024 02:44:59.214Z][Thread 0023][akka://System/user/41] 5065
[INFO][04/11/2024 02:44:59.214Z][Thread 0027][akka://System/user/42] 1516
[INFO][04/11/2024 02:44:59.214Z][Thread 0028][akka://System/user/7] 509
[INFO][04/11/2024 02:44:59.214Z][Thread 0012][akka://System/user/4] 6345
[INFO][04/11/2024 02:44:59.214Z][Thread 0015][akka://System/user/35] 6345
[INFO][04/11/2024 02:44:59.214Z][Thread 0014][akka://System/user/86] 6344
[INFO][04/11/2024 02:44:59.214Z][Thread 0025][akka://System/user/4999] 3014
[INFO][04/11/2024 02:44:59.214Z][Thread 0019][akka://System/user/38] 6344
[INFO][04/11/2024 02:44:59.239Z][Thread 0021][akka://System/user/43] 0
[INFO][04/11/2024 02:44:59.239Z][Thread 0014][akka://System/user/85] 0
[INFO][04/11/2024 02:44:59.239Z][Thread 0020][akka://System/user/10] 0
```

I'm not getting any useful contention data back using `dotnet-counters`, but I do see deadlocks at startup occasionally - which @Arkatufus was also able to reproduce. I'm going to test this with his fix real quick.

Looks like @Arkatufus ' solution on #7144 fixes this issue:

## NET 8.0

```text
C:\Program Files\dotnet\sdk\8.0.104\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.EolTargetFrameworks.targets(32,5): warning NETSDK1138: The target framework 'netcoreapp3.1' is out of support and will not receive security updates in the future. Please refer to https://aka.ms/dotnet-core-support for more information about the support policy. [E:\Repositories\repros\HighSchedulerCpu\HighSchedulerCpu.csproj::TargetFramework=netcoreapp3.1]
E:\Repositories\repros\HighSchedulerCpu\Program.cs(29,9): warning CS8602: Dereference of a possibly null reference. [E:\Repositories\repros\HighSchedulerCpu\HighSchedulerCpu.csproj::TargetFramework=net8.0]
Waiting to start
Starting
[INFO][04/11/2024 03:08:48.715Z][Thread 0025][akka://System/user/12] 2
[INFO][04/11/2024 03:08:48.716Z][Thread 0009][akka://System/user/2] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0018][akka://System/user/9] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0030][akka://System/user/81] 1
[INFO][04/11/2024 03:08:48.716Z][Thread 0027][akka://System/user/80] 2
[INFO][04/11/2024 03:08:48.716Z][Thread 0007][akka://System/user/1] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0022][akka://System/user/78] 2
[INFO][04/11/2024 03:08:48.716Z][Thread 0036][akka://System/user/738] 0
[INFO][04/11/2024 03:08:48.716Z][Thread 0033][akka://System/user/82] 1
[INFO][04/11/2024 03:08:48.716Z][Thread 0017][akka://System/user/148] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0013][akka://System/user/5] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0034][akka://System/user/185] 1
[INFO][04/11/2024 03:08:48.716Z][Thread 0020][akka://System/user/182] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0037][akka://System/user/15] 0
[INFO][04/11/2024 03:08:48.716Z][Thread 0023][akka://System/user/11] 2
[INFO][04/11/2024 03:08:48.716Z][Thread 0031][akka://System/user/184] 1
[INFO][04/11/2024 03:08:48.716Z][Thread 0035][akka://System/user/14] 0
[INFO][04/11/2024 03:08:48.716Z][Thread 0032][akka://System/user/13] 1
[INFO][04/11/2024 03:08:48.716Z][Thread 0014][akka://System/user/6] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0011][akka://System/user/3] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0015][akka://System/user/7] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0021][akka://System/user/149] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0024][akka://System/user/79] 2
[INFO][04/11/2024 03:08:48.716Z][Thread 0008][akka://System/user/0] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0010][akka://System/user/4] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0028][akka://System/user/183] 1
[INFO][04/11/2024 03:08:48.716Z][Thread 0012][akka://System/user/77] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0026][akka://System/user/708] 0
[INFO][04/11/2024 03:08:48.716Z][Thread 0019][akka://System/user/10] 9
[INFO][04/11/2024 03:08:48.716Z][Thread 0016][akka://System/user/8] 9
[INFO][04/11/2024 03:08:48.751Z][Thread 0018][akka://System/user/707] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0032][akka://System/user/736] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0011][akka://System/user/706] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0033][akka://System/user/733] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0013][akka://System/user/737] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0007][akka://System/user/734] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0021][akka://System/user/735] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0033][akka://System/user/730] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0007][akka://System/user/729] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0013][akka://System/user/700] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0010][akka://System/user/732] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0033][akka://System/user/728] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0007][akka://System/user/727] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0013][akka://System/user/699] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0017][akka://System/user/147] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0007][akka://System/user/697] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0013][akka://System/user/724] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0033][akka://System/user/698] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0010][akka://System/user/725] 0
[INFO][04/11/2024 03:08:48.751Z][Thread 0030][akka://System/user/705] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0021][akka://System/user/726] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0010][akka://System/user/695] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0033][akka://System/user/723] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0013][akka://System/user/696] 0
[INFO][04/11/2024 03:08:48.752Z][Thread 0007][akka://System/user/722] 0
```

--- a/src/core/Akka.Tests/Actor/Scheduler/HashedWheelTimerSchedulerContentionSpec.cs
+++ b/src/core/Akka.Tests/Actor/Scheduler/HashedWheelTimerSchedulerContentionSpec.cs
@@ -32,25 +32,15 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
     [Fact]
     public void SchedulerContentionTest()
     {
-        var collector = CreateTestProbe();
-        
         foreach (var i in Enumerable.Range(0, TotalActor))
         {
-            Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor, collector)), i.ToString());
+            Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor)), i.ToString());
         }
 
-        Within(10.Seconds(), () =>
-        {
-            for (var i = 0; i < TotalActor; i++)
-            {
-                ExpectMsg<Done>();
-            }
-        });
-        
         object? received = null;
         do
         {
-            received = collector.ReceiveOne(TimeSpan.Zero);
+            received = ReceiveOne(TimeSpan.Zero);
             if (received is long value)
             {
                 value.Should().BeLessThan(200, "Scheduler should not experience resource contention");
@@ -62,7 +52,6 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
     [Fact]
     public void SchedulerContentionThreadedTest()
     {
-        var collector = CreateTestProbe();
         var threads = new List<Thread>();
         
         foreach (var j in Enumerable.Range(0, TotalThreads))
@@ -80,18 +69,10 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
             thread.Join();
         }
 
-        Within(10.Seconds(), () =>
-        {
-            for (var i = 0; i < TotalActor; i++)
-            {
-                ExpectMsg<Done>();
-            }
-        });
-
         object? received = null;
         do
         {
-            received = collector.ReceiveOne(TimeSpan.Zero);
+            received = ReceiveOne(TimeSpan.Zero);
             if (received is long value)
             {
                 value.Should().BeLessThan(200, "Scheduler should not experience resource contention");
@@ -105,7 +86,7 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
             n *= ActorsPerThread;
             for (var i = 0; i < ActorsPerThread; i++)
             {
-                Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor, collector)), (n + i).ToString());
+                Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor)), (n + i).ToString());
             }
         }
     }
@@ -114,12 +95,11 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
     {
         public ITimerScheduler Timers { get; set; }
 
-        public DoStuffActor(IActorRef probe, IActorRef collector)
+        public DoStuffActor(IActorRef probe)
         {
             Receive<Done>(d =>
             {
                 Context.Stop(Self);
-                probe.Tell(d);
             });
 
             var sw = Stopwatch.StartNew();
@@ -129,7 +109,7 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
             if (sw.ElapsedMilliseconds > 0)
             {
                 Context.GetLogger().Info($"{sw.ElapsedMilliseconds}");
-                collector.Tell(sw.ElapsedMilliseconds);
+                probe.Tell(sw.ElapsedMilliseconds);
             }
         }
     }

--- a/src/core/Akka.Tests/Actor/Scheduler/HashedWheelTimerSchedulerContentionSpec.cs
+++ b/src/core/Akka.Tests/Actor/Scheduler/HashedWheelTimerSchedulerContentionSpec.cs
@@ -32,15 +32,24 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
     [Fact]
     public void SchedulerContentionTest()
     {
+        var collector = CreateTestProbe();
         foreach (var i in Enumerable.Range(0, TotalActor))
         {
-            Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor)), i.ToString());
+            Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor, collector)), i.ToString());
         }
+
+        Within(10.Seconds(), () =>
+        {
+            for (var x = 0; x < TotalActor; x++)
+            {
+                ExpectMsg<Done>();
+            }
+        });
 
         object? received = null;
         do
         {
-            received = ReceiveOne(TimeSpan.Zero);
+            received = collector.ReceiveOne(TimeSpan.Zero);
             if (received is long value)
             {
                 value.Should().BeLessThan(200, "Scheduler should not experience resource contention");
@@ -52,6 +61,7 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
     [Fact]
     public void SchedulerContentionThreadedTest()
     {
+        var collector = CreateTestProbe();
         var threads = new List<Thread>();
         
         foreach (var j in Enumerable.Range(0, TotalThreads))
@@ -69,10 +79,18 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
             thread.Join();
         }
 
+        Within(10.Seconds(), () =>
+        {
+            for (var x = 0; x < TotalActor; x++)
+            {
+                ExpectMsg<Done>();
+            }
+        });
+
         object? received = null;
         do
         {
-            received = ReceiveOne(TimeSpan.Zero);
+            received = collector.ReceiveOne(TimeSpan.Zero);
             if (received is long value)
             {
                 value.Should().BeLessThan(200, "Scheduler should not experience resource contention");
@@ -86,22 +104,30 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
             n *= ActorsPerThread;
             for (var i = 0; i < ActorsPerThread; i++)
             {
-                Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor)), (n + i).ToString());
+                Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor, collector)), (n + i).ToString());
             }
         }
     }
     
     public class DoStuffActor : ReceiveActor, IWithTimers
     {
+        private readonly IActorRef _collector;
         public ITimerScheduler Timers { get; set; }
 
-        public DoStuffActor(IActorRef probe)
+        public DoStuffActor(IActorRef probe, IActorRef collector)
         {
+            _collector = collector;
+            
             Receive<Done>(d =>
             {
                 Context.Stop(Self);
+                probe.Tell(d);
             });
+        }
 
+        protected override void PreStart()
+        {
+            base.PreStart();
             var sw = Stopwatch.StartNew();
             Timers.StartSingleTimer("Test", Done.Instance, TimeSpan.FromSeconds(3));
             sw.Stop();
@@ -109,7 +135,7 @@ public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
             if (sw.ElapsedMilliseconds > 0)
             {
                 Context.GetLogger().Info($"{sw.ElapsedMilliseconds}");
-                probe.Tell(sw.ElapsedMilliseconds);
+                _collector.Tell(sw.ElapsedMilliseconds);
             }
         }
     }

--- a/src/core/Akka.Tests/Actor/Scheduler/HashedWheelTimerSchedulerContentionSpec.cs
+++ b/src/core/Akka.Tests/Actor/Scheduler/HashedWheelTimerSchedulerContentionSpec.cs
@@ -1,0 +1,136 @@
+﻿// -----------------------------------------------------------------------
+//  <copyright file="SchedulerHeavyUse.cs" company="Akka.NET Project">
+//      Copyright (C) 2009-2024 Lightbend Inc. <http://www.lightbend.com>
+//      Copyright (C) 2013-2024 .NET Foundation <https://github.com/akkadotnet/akka.net>
+//  </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Akka.Actor;
+using Akka.Event;
+using FluentAssertions;
+using FluentAssertions.Extensions;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Akka.Tests.Actor.Scheduler;
+
+public class HashedWheelTimerSchedulerContentionSpec: TestKit.Xunit2.TestKit
+{
+    private const int TotalActor = 5000;
+    private const int TotalThreads = 10;
+    private const int ActorsPerThread = TotalActor / TotalThreads;
+    
+    public HashedWheelTimerSchedulerContentionSpec(ITestOutputHelper output) : base("{}", output)
+    {
+    }
+
+    [Fact]
+    public void SchedulerContentionTest()
+    {
+        var collector = CreateTestProbe();
+        
+        foreach (var i in Enumerable.Range(0, TotalActor))
+        {
+            Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor, collector)), i.ToString());
+        }
+
+        Within(10.Seconds(), () =>
+        {
+            for (var i = 0; i < TotalActor; i++)
+            {
+                ExpectMsg<Done>();
+            }
+        });
+        
+        object? received = null;
+        do
+        {
+            received = collector.ReceiveOne(TimeSpan.Zero);
+            if (received is long value)
+            {
+                value.Should().BeLessThan(200, "Scheduler should not experience resource contention");
+            }
+        } while (received is not null);
+        
+    }
+    
+    [Fact]
+    public void SchedulerContentionThreadedTest()
+    {
+        var collector = CreateTestProbe();
+        var threads = new List<Thread>();
+        
+        foreach (var j in Enumerable.Range(0, TotalThreads))
+        {
+            threads.Add(new Thread(() => RunThread(j)));
+        }
+
+        foreach (var thread in threads)
+        {
+            thread.Start();
+        }
+
+        foreach (var thread in threads)
+        {
+            thread.Join();
+        }
+
+        Within(10.Seconds(), () =>
+        {
+            for (var i = 0; i < TotalActor; i++)
+            {
+                ExpectMsg<Done>();
+            }
+        });
+
+        object? received = null;
+        do
+        {
+            received = collector.ReceiveOne(TimeSpan.Zero);
+            if (received is long value)
+            {
+                value.Should().BeLessThan(200, "Scheduler should not experience resource contention");
+            }
+        } while (received is not null);
+
+        return;
+
+        void RunThread(int n)
+        {
+            n *= ActorsPerThread;
+            for (var i = 0; i < ActorsPerThread; i++)
+            {
+                Sys.ActorOf(Props.Create(() => new DoStuffActor(TestActor, collector)), (n + i).ToString());
+            }
+        }
+    }
+    
+    public class DoStuffActor : ReceiveActor, IWithTimers
+    {
+        public ITimerScheduler Timers { get; set; }
+
+        public DoStuffActor(IActorRef probe, IActorRef collector)
+        {
+            Receive<Done>(d =>
+            {
+                Context.Stop(Self);
+                probe.Tell(d);
+            });
+
+            var sw = Stopwatch.StartNew();
+            Timers.StartSingleTimer("Test", Done.Instance, TimeSpan.FromSeconds(3));
+            sw.Stop();
+
+            if (sw.ElapsedMilliseconds > 0)
+            {
+                Context.GetLogger().Info($"{sw.ElapsedMilliseconds}");
+                collector.Tell(sw.ElapsedMilliseconds);
+            }
+        }
+    }
+}

--- a/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
+++ b/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
@@ -147,24 +147,25 @@ namespace Akka.Actor
 
         private void Start()
         {
-            if (_workerState != WORKER_STATE_STARTED)
+            if (_workerState == WORKER_STATE_STARTED)
             {
-                if (_workerState == WORKER_STATE_INIT)
+                // do nothing
+            }
+            else if (_workerState == WORKER_STATE_INIT)
+            {
+                if (Interlocked.CompareExchange(ref _workerState, WORKER_STATE_STARTED, WORKER_STATE_INIT) == WORKER_STATE_INIT)
                 {
-                    if (Interlocked.CompareExchange(ref _workerState, WORKER_STATE_STARTED, WORKER_STATE_INIT) == WORKER_STATE_INIT)
-                    {
-                        _timer ??= new PeriodicTimer(_timerDuration);
-                        Task.Run(() => RunAsync(_cts.Token)); // start the clock
-                    }
+                    _timer ??= new PeriodicTimer(_timerDuration);
+                    Task.Run(() => RunAsync(_cts.Token)); // start the clock
                 }
-                else if (_workerState == WORKER_STATE_SHUTDOWN)
-                {
-                    throw new SchedulerException("cannot enqueue after timer shutdown");
-                }
-                else
-                {
-                    throw new InvalidOperationException($"Worker in invalid state: {_workerState}");
-                }
+            }
+            else if (_workerState == WORKER_STATE_SHUTDOWN)
+            {
+                throw new SchedulerException("cannot enqueue after timer shutdown");
+            }
+            else
+            {
+                throw new InvalidOperationException($"Worker in invalid state: {_workerState}");
             }
 
             if(_startTime == 0)

--- a/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
+++ b/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
@@ -34,6 +34,7 @@ namespace Akka.Actor
     public sealed class HashedWheelTimerScheduler : SchedulerBase, IDateTimeOffsetNowTimeProvider, IDisposable
     {
         private readonly TimeSpan _shutdownTimeout;
+        private readonly TimeSpan _timerDuration;
         private readonly long _tickDuration; // a timespan expressed as ticks
 
         static HashedWheelTimerScheduler()
@@ -65,17 +66,17 @@ namespace Akka.Actor
                     "Cannot be greater than 2^30.");
             // ReSharper restore NotResolvedInText
             
-            var tickDuration = SchedulerConfig.GetTimeSpan("akka.scheduler.tick-duration", TimeSpan.Zero);
-            if (tickDuration.TotalMilliseconds < 10.0d)
+            _timerDuration = SchedulerConfig.GetTimeSpan("akka.scheduler.tick-duration", TimeSpan.Zero);
+            if (_timerDuration.TotalMilliseconds < 10.0d)
             {
                 // ReSharper disable NotResolvedInText
-                throw new ArgumentOutOfRangeException("akka.scheduler.tick-duration", tickDuration.TotalMilliseconds,
+                throw new ArgumentOutOfRangeException("akka.scheduler.tick-duration", _timerDuration.TotalMilliseconds,
                     "minimum supported akka.scheduler.tick-duration on Windows is 10ms");
                 // ReSharper restore NotResolvedInText
             }
 
             // convert tick-duration to ticks
-            _tickDuration = tickDuration.Ticks;
+            _tickDuration = _timerDuration.Ticks;
 
             // Normalize ticks per wheel to power of two and create the wheel
             _wheel = CreateWheel(ticksPerWheel, log);
@@ -96,7 +97,11 @@ namespace Akka.Actor
         private long _startTime;
         private long _tick;
         private readonly int _mask;
+#if NET6_0_OR_GREATER
+        private readonly TaskCompletionSource _workerInitialized = new();
+#else
         private readonly CountdownEvent _workerInitialized = new(1);
+#endif
         private readonly ConcurrentQueue<SchedulerRegistration> _registrations = new();
         private readonly Bucket[] _wheel;
 
@@ -137,24 +142,21 @@ namespace Akka.Actor
         private readonly HashSet<SchedulerRegistration> _rescheduleRegistrations = new();
         
 #if NET6_0_OR_GREATER
-        private readonly object _lock = new ();
         private PeriodicTimer? _timer;
         private readonly CancellationTokenSource _cts = new();
 
         private void Start()
         {
-            var start = false;
-            lock (_lock)
+            if (_workerState != WORKER_STATE_STARTED)
             {
                 switch (_workerState)
                 {
-                    case WORKER_STATE_STARTED:
-                        // do nothing
-                        break;
-                
                     case WORKER_STATE_INIT:
-                        _workerState = WORKER_STATE_STARTED;
-                        start = true;
+                        if (Interlocked.CompareExchange(ref _workerState, WORKER_STATE_STARTED, WORKER_STATE_INIT) == WORKER_STATE_INIT)
+                        {
+                            _timer ??= new PeriodicTimer(_timerDuration);
+                            Task.Run(() => RunAsync(_cts.Token)); // start the clock
+                        }
                         break;
                 
                     case WORKER_STATE_SHUTDOWN:
@@ -165,16 +167,8 @@ namespace Akka.Actor
                 }
             }
 
-            if (start)
-            {
-                var timerDuration = TimeSpan.FromTicks(_tickDuration);
-                _timer ??= new PeriodicTimer(timerDuration);
-                Task.Run(() => RunAsync(_cts.Token)); // start the clock
-            }
-            else
-            {
-                _workerInitialized.Wait();
-            }
+            if(_startTime == 0)
+                _workerInitialized.Task.Wait();
         }
 
         private async Task RunAsync(CancellationToken token)
@@ -190,7 +184,7 @@ namespace Akka.Actor
                 _startTime = 1;
             }
 
-            _workerInitialized.Signal();
+            _workerInitialized.SetResult();
 
             try
             {

--- a/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
+++ b/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
@@ -137,35 +137,41 @@ namespace Akka.Actor
         private readonly HashSet<SchedulerRegistration> _rescheduleRegistrations = new();
         
 #if NET6_0_OR_GREATER
+        private readonly object _lock = new ();
         private PeriodicTimer? _timer;
         private readonly CancellationTokenSource _cts = new();
 
         private void Start()
         {
-            if (_workerState == WORKER_STATE_STARTED)
+            var start = false;
+            lock (_lock)
             {
-            } // do nothing
-            else if (_workerState == WORKER_STATE_INIT)
-            {
-                if (Interlocked.CompareExchange(ref _workerState, WORKER_STATE_STARTED, WORKER_STATE_INIT) ==
-                    WORKER_STATE_INIT)
+                switch (_workerState)
                 {
-                    var t = TimeSpan.FromTicks(_tickDuration);
-                    _timer = new PeriodicTimer(t);
-                    
-                    Task.Run(() => RunAsync(_cts.Token)); // start the clock
+                    case WORKER_STATE_STARTED:
+                        // do nothing
+                        break;
+                
+                    case WORKER_STATE_INIT:
+                        _workerState = WORKER_STATE_STARTED;
+                        start = true;
+                        break;
+                
+                    case WORKER_STATE_SHUTDOWN:
+                        throw new SchedulerException("cannot enqueue after timer shutdown");
+                
+                    default:
+                        throw new InvalidOperationException($"Worker in invalid state: {_workerState}");
                 }
             }
-            else if (_workerState == WORKER_STATE_SHUTDOWN)
+
+            if (start)
             {
-                throw new SchedulerException("cannot enqueue after timer shutdown");
+                var timerDuration = TimeSpan.FromTicks(_tickDuration);
+                _timer ??= new PeriodicTimer(timerDuration);
+                Task.Run(() => RunAsync(_cts.Token)); // start the clock
             }
             else
-            {
-                throw new InvalidOperationException($"Worker in invalid state: {_workerState}");
-            }
-                    
-            while (_startTime == 0)
             {
                 _workerInitialized.Wait();
             }
@@ -248,7 +254,7 @@ namespace Akka.Actor
             if (_workerState == WORKER_STATE_STARTED) { } // do nothing
             else if (_workerState == WORKER_STATE_INIT)
             {
-                _worker = new Thread(Run) { IsBackground = true };
+                _worker ??= new Thread(Run) { IsBackground = true };
                 if (Interlocked.CompareExchange(ref _workerState, WORKER_STATE_STARTED, WORKER_STATE_INIT) ==
                     WORKER_STATE_INIT)
                 {

--- a/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
+++ b/src/core/Akka/Actor/Scheduler/HashedWheelTimerScheduler.cs
@@ -149,21 +149,21 @@ namespace Akka.Actor
         {
             if (_workerState != WORKER_STATE_STARTED)
             {
-                switch (_workerState)
+                if (_workerState == WORKER_STATE_INIT)
                 {
-                    case WORKER_STATE_INIT:
-                        if (Interlocked.CompareExchange(ref _workerState, WORKER_STATE_STARTED, WORKER_STATE_INIT) == WORKER_STATE_INIT)
-                        {
-                            _timer ??= new PeriodicTimer(_timerDuration);
-                            Task.Run(() => RunAsync(_cts.Token)); // start the clock
-                        }
-                        break;
-                
-                    case WORKER_STATE_SHUTDOWN:
-                        throw new SchedulerException("cannot enqueue after timer shutdown");
-                
-                    default:
-                        throw new InvalidOperationException($"Worker in invalid state: {_workerState}");
+                    if (Interlocked.CompareExchange(ref _workerState, WORKER_STATE_STARTED, WORKER_STATE_INIT) == WORKER_STATE_INIT)
+                    {
+                        _timer ??= new PeriodicTimer(_timerDuration);
+                        Task.Run(() => RunAsync(_cts.Token)); // start the clock
+                    }
+                }
+                else if (_workerState == WORKER_STATE_SHUTDOWN)
+                {
+                    throw new SchedulerException("cannot enqueue after timer shutdown");
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Worker in invalid state: {_workerState}");
                 }
             }
 

